slides/02_traits.md

```diff
@@ -0,0 +1,376 @@
+---
+jupyter:
+  jupytext:
+    text_representation:
+      extension: .md
+      format_name: markdown
+      format_version: '1.3'
+      jupytext_version: 1.13.7
+  kernelspec:
+    display_name: Python 3 (ipykernel)
+    language: python
+    name: python3
+---
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Step1: Using Traits
+
+**Jonathan Rocher, Siddhant Wahal, Jason Chambless, Prabhu Ramachandran**
+
+**SciPy 2022**
+
+<!-- #endregion -->
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Introduction to Traits
+
+- **trait**: Python object attribute with additional characteristics
+    - Typed attributes
+    - Reactive
+    - Observable
+    - Cleaner code
+    - Easy UI
+
+<br/>
+
+- https://docs.enthought.com/traits/
+- https://github.com/enthought/traits/
+
+<!-- #endregion -->
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Why all this?
+
+- No pain, no gain!
+
+- Small change to thinking yields big benefits
+
+- Do not mix GUI code with core model
+
+- Build a clean model first
+
+  - Easier to understand/maintain
+  - Separation of concerns
+  - Easier to test
+  - Generally better reuse
+
+<!-- #endregion -->
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Trait features
+
+- Initialization: default value
+- Validation: strongly typed
+- Deferral/Delegation: value delegation
+- Notification: events
+- Visualization: MVC, automatic GUI!
+
+<!-- #endregion -->
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## An example
+
+<!-- #endregion -->
+
+```python
+from traits.api import Delegate, HasStrictTraits, Instance, Int, Str, observe
+
+class Parent(HasStrictTraits):
+    # INITIALIZATION: 'last_name' initialized to ''
+    last_name = Str('')
+
+```
+
+```python
+class Child(HasStrictTraits):
+    age = Int
+    # VALIDATION: 'father' must be Parent instance
+    father = Instance(Parent)
+    # DELEGATION: 'last_name' delegated to father's
+    last_name = Delegate('father')
+    # NOTIFICATION: Method called when 'age' changes
+    def _age_changed(self, old, new):
```

Comment:
Statically defining listeners with the `_trait_changed` magic method is now deprecated. See https://docs.enthought.com/traits/traits_user_manual/listening.html#on-trait-change-notification

Reply:
Oh, that is a huge change -- there is a ton of code that uses that! I don't see a specific deprecation.  What is deprecated is using on_trait_change, I thought.  I searched for "deprecate" but found no mention.

Reply:
Sorry, I was a bit loose with my language. It's not deprecated, but could be in the future.

https://docs.enthought.com/traits/traits_user_manual/notification.html#trait-notification

> Traits 6.1 introduces a new API for configuring traits notifications: observe, which is intended to fully replace an older API (on_trait_change) in order to overcome some of its limitations and flaws. While on_trait_change is still supported, it may be removed in the future.

My understanding is that `observe` is now the recommended way of listening to trait changes.

Reply:
I think the static handlers are too widely used to be thrown away and they are also not too hard -- for one thing, I find that the forced naming makes it easier to use an appropriate method name. I will leave it there in the slides for now.